rust_icu_uloc/src/lib.rs

```diff
@@ -346,43 +496,95 @@ pub fn accept_language(
         .map(|uloc| (Some(uloc), accept_result))
 }
 
+/// Implements `uloc_toUnicodeLocaleKey` from ICU4C.
+pub fn to_unicode_locale_key(legacy_keyword: &str) -> Option<String> {
+    let legacy_keyword = str_to_cstring(legacy_keyword).unwrap();
+    let unicode_keyword: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toUnicodeLocaleKey)(legacy_keyword.as_ptr());
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    unicode_keyword.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+/// Implements `uloc_toUnicodeLocaleType` from ICU4C.
+pub fn to_unicode_locale_type(legacy_keyword: &str, legacy_value: &str) -> Option<String> {
+    let legacy_keyword = str_to_cstring(legacy_keyword).unwrap();
+    let legacy_value = str_to_cstring(legacy_value).unwrap();
+    let unicode_value: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toUnicodeLocaleType)(
+            legacy_keyword.as_ptr(),
+            legacy_value.as_ptr(),
+        );
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    unicode_value.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+/// Implements `uloc_toLegacyKey` from ICU4C.
+pub fn to_legacy_key(unicode_keyword: &str) -> Option<String> {
+    let unicode_keyword = str_to_cstring(unicode_keyword).unwrap();
+    let legacy_keyword: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toLegacyKey)(unicode_keyword.as_ptr());
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    legacy_keyword.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+fn str_to_cstring(input: &str) -> Result<ffi::CString, common::Error> {
+    ffi::CString::new(input).map_err(|_| common::Error::string_with_interior_nul())
+}
+
+fn cstring_to_string(input: &ffi::CString) -> Result<String, common::Error> {
+    input
+        .to_str()
+        .map(|s| s.to_owned())
+        .map_err(|e| common::Error::Wrapper(e.to_string().into()))
+}
+
 #[cfg(test)]
 mod tests {
-    use super::*;
+    use {super::*, anyhow::Error};
+
+    #[test]
+    fn test_language() -> Result<(), Error> {
```

Comment:
Do these tests need to return a `Result` now that we've simplified the type signatures of the getters?

Reply:
All the `ULoc` factory methods still return `Result`s, and I find using `?` cleaner than having to write explicit `.expect("this should never fail")` (or just `.unwrap()`) calls in unit tests.